pkg/pillar/cmd/zedagent/zedagent.go

```diff
@@ -287,23 +287,32 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
 	// Initialize all zedagent publications.
 	initPublications(zedagentCtx)
 
-	// upgradeconverter ensures we have a ConfigItemValueMap so we
-	// read it to get the initial values
-	item, err := zedagentCtx.pubGlobalConfig.Get("global")
-	if err != nil {
-		log.Fatalf("ConfigItemValueMap missing: %s", err)
-	}
-	zedagentCtx.globalConfig = item.(types.ConfigItemValueMap)
-	log.Functionf("initialized GlobalConfig")
-
 	// Run a periodic timer so we always update StillRunning
 	stillRunning := time.NewTicker(25 * time.Second)
 	ps.StillRunning(agentName, warningTime, errorTime)
 
 	initializeDirs()
 
-	// Apply saved radio config ASAP.
+	// Load bootstrap configuration if present.
 	getconfigCtx := zedagentCtx.getconfigCtx
+	loadBootstrapConfig(getconfigCtx)
+
+	// Get GlobalConfig.
+	// If not present (e.g. loading of bootstrap config failed), use default values.
```

Comment:
How does this interact with upgradeconverter?

We need the logic in upgradeconverter which ensures that when a new item has been added in the new version of EVE-OS, it gets filled in with its default value.

Can't we have loadBootStrapConfig override all of configItemValueMap instead?

Reply:
Let me explain the current workflow:

1. upgradeconverter runs [importFromConfigPartition](https://github.com/milan-zededa/eve/blob/bootstrap-config/pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go#L80) (link to my fork with these changes included). If /`/config/bootstrap-config.pb` is NOT present, it will act as before, i.e. copying global.json from `/config` to `/persist/zedagent/ConfigItemValueMap/`. If bootstrap config is present, this step is skipped.
2. upgradeconverter runs [applyDefaultConfigItem](https://github.com/milan-zededa/eve/blob/bootstrap-config/pkg/pillar/cmd/upgradeconverter/applydefaultconfigitem.go#L27). If global config is published (from previous boot), it just adds default values of missing keys (Eve was possibly upgraded and the new version has new config items). This is done regardless if there is any bootstrap config (i.e. the behavior is unchanged). If global config is missing (not published, likely the very first boot), then default values are published if bootstrap config is NOT present.
3. zedagent tries to load bootstrap config if present. If successful, config items (aka global config) are published. If not successful, zedagent publishes default values (with all keys known to the running EVE version).

So basically the goal was to prefer bootstrap config and publish default values only as a last resort. We could also make it so that default values are published by upgradeconverter and then possible get overwritten by zedagent during bootstrap loading, but there would be a small window during which the default config would be in effect, which could cause some confusion I think.
But in all those possible code-paths it is ensured that published global config has all the keys known to running EVE included, which was you main worry.

Reply:
In step 2 doesn't upgradeconverter always end up publishing the configItemValueMap? Or are you changing that?
I don't know what other parts of the system assume that there is a configItemValueMap...

Reply:
Resolved in the discussion below.